lib/common/socket/uv-binding.c.h

```diff
@@ -197,7 +197,10 @@ void do_write(h2o_socket_t *_sock, h2o_iovec_t *bufs, size_t bufcnt, h2o_socket_
     assert(sock->super._cb.write == NULL);
     sock->super._cb.write = cb;
 
-    uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, (uv_buf_t *)bufs, (int)bufcnt, on_do_write_complete);
+    if (bufcnt > 0)
+        uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, (uv_buf_t *)bufs, (int)bufcnt, on_do_write_complete);
+    else
+        on_do_write_complete(&sock->stream._wreq, 0);
```

Comment:
I think we have to invoke `on_do_write_complete` asynchronously. Callers of `h2o_socket_write` do not expect the completion callback to be invoked while `h2o_socket_write` is running.

To defer the invocation, I think the easiest way would be to add a `h2o_timer_t` instance to `st_h2o_uv_socket_t` and do `h2o_timer_link` with a timeout value of zero.